### PR TITLE
Safe precompile methods (#854)

### PR DESCRIPTION
**Pull request by stoqnkpL**
This PR introduces the `SafeHTS` Solidity library which contains safe methods for the HTS precompiled contract. Users can use this library if they want the messageFrame to revert and execution of their smart contracts to halt if the precompiled contract message call is not successful.

Opened from https://github.com/hashgraph/hedera-services/pull/2781 with squashed changes (git wouldn't allow me to force push in the old branch)

### CODE REVIEW

File: test-clients/src/main/resource/contract/solidity/hip-206/safe/SafeHTS.sol [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    function safeTransferNFT(IHTS token, address sender, address receiver, int64 serialNumber) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHTS.transferNFT.selector,
+            token, sender, receiver, serialNumber));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe NFT transfer failed!");
+    }
+
+    function safeCreateFungibleToken(IHTS.HederaToken memory token, uint initialTotalSupply,
```
Comment (stoqnkpL) [inline test-clients/src/main/resource/contract/solidity/hip-206/safe/SafeHTS.sol:119]: Can we add the [modifier](https://github.com/LimeChain/hedera-smart-contracts/blob/b1a7f33a4f49fbb2c0134f1420ffaec0d0a4ec06/hts-precompile/HederaTokenService.sol#L22) for default autoRenewPeriod to this library? This way we would be able to call the create functions with minimal required fields.

File: test-clients/src/main/resource/contract/bytecodes/SafeOperationsContract.bin [hunk 2]
```diff
@@ -1 +1 @@
...
...6082015260800190565b6020808252601c908201527f5361666520746f6b656e73207472616e73666572206661696c65642100000000604082015260600190565b60208082526011908201527053616665206275726e206661696c65642160781b604082015260600190565b60208082526019908201527f53616665204e4654207472616e73666572206661696c65642100000000000000604082015260600190565b6020808252601f908201527f536166652073696e676c65206173736f63696174696f6e206661696c65642100604082015260600190565b6020808252818101527f536166652073696e676c6520646973736f63696174696f6e206661696c656421604082015260600190565b60208082526011908201527053616665206d696e74206661696c65642160781b604082015260600190565b6001600160401b0391909116815260200190565b60006001600160401b03841682526040602083015261025c60408301846115c4565b6040518181016001600160401b0381118282101715611bca57600080fd5b604052919050565b60006001600160401b03821115611be7578081fd5b5060209081020190565b90815260200190565b60005b83811015611c15578181015183820152602001611bfd565b838111156102185750506000910152565b6001600160a01b038116811461026d57600080fd5b8060070b811461026d57600080fd5b6001600160401b038116811461026d57600080fdfea26469706673582212209ab7df136a412e0365ea8696787884f84e59110719286c88d35424f3644099b764736f6c634300060c0033
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/resource/contract/bytecodes/SafeOperationsContract.bin:1]: This file would soon be obsolete. Only the .bin file under contract/contracts/SafeOperations will be used.

Comment (stoqnkpL) [reply test-clients/src/main/resource/contract/bytecodes/SafeOperationsContract.bin:1]: I suggest merging branch `fix-remaining-eth-isomers-delete-contract-resources` from https://github.com/hashgraph/hedera-services/pull/3465 into `safe-precompile-methods`.

**Comment by stoqnkpL** [comment]
There's a failing test in `SafeOperationsSuite`: safeMultipleAssociatePrecompile()

**Comment by stoqnkpL** [comment]
The Solidity code for `SafeHTS` looks good to me.

File: test-clients/src/main/resource/contract/solidity/hip-206/safe/SafeHTS.sol [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHTS.cryptoTransfer.selector, tokenTransfers));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe crypto transfer failed!");
+    }
+
+    function safeMintToken(IHTS token, uint64 amount, bytes[] memory metadata) internal
+    returns (uint64 newTotalSupply, int64[] memory serialNumbers) {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.delegatecall(
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/resource/contract/solidity/hip-206/safe/SafeHTS.sol:25]: Do we need `delegatecall` here?

**Review by IvanKavaldzhiev — COMMENTED** [review]
LGTM. I think we only should resolve the issue with `delegatecall`

**Comment by shemnon** [comment]
I think Hedera Services is the wrong place for this contract to live in and to be tested.  I think that https://github.com/hashgraph/hedera-smart-contracts/ is the correct repo for this lib to live in, and a testing app/framework should be set up that can test local or public networks (ideally only testnet and previewnet) should be added to verify the safe variants work as expected.

**Comment by Nana-EC** [comment]
@stoqnkpL 
Please close this PR and add a PR on the [Smart Contracts Repo](https://github.com/hashgraph/hedera-smart-contracts/) as Danno noted.
You can add it under the hts-precompile folder and open a ticket there that will address adding test support for it in the future.

**Comment by stoqnkpL** [comment]
Closing this PR, opened https://github.com/hashgraph/hedera-smart-contracts/pull/41 adding the SafeHTS library to the smart-contracts repo. Also opened https://github.com/hashgraph/hedera-smart-contracts/issues/42 for future work on SafeHTS testing.

**Pull request closed without merging**